engine/verification/test/assignerfetcher_test.go

```diff
@@ -0,0 +1,291 @@
+// block reader receives (container) finalized blocks that contain execution receipts preceding (reference) blocks.
+// some receipts have duplicate results.
+// - in a staked verification node:
+// -- for each distinct result assigner engine receives, it does the chunk assignment and passes the
+// chunk locators of assigned chunks to chunk queue, which in turn delivers to chunks processor though the chunks consumer.
+// - in an unstaked verification node:
+// -- execution results are discarded.
+// - it does a correct resource clean up of the pipeline after handling all incoming receipts
+func TestAssignerFetcherPipeline(t *testing.T) {
+	testcases := []struct {
+		blockCount int
+		ops        []utils.CompleteExecutionReceiptBuilderOpt
+		msg        string
+		staked     bool
+	}{
+		{
+			// read this test case in this way:
+			// one block is passed to block reader. The block contains one
+			// execution result that is not duplicate (single copy).
+			// The result has only one chunk.
+			// The verification node is staked
+			blockCount: 1,
+			ops: []utils.CompleteExecutionReceiptBuilderOpt{
+				utils.WithResults(1),
+				utils.WithChunks(1),
+				utils.WithCopies(1),
+			},
+			staked: true,
+			msg:    "1 block, 1 result, 1 chunk, no duplicate, staked",
+		},
+		{
+			blockCount: 1,
+			ops: []utils.CompleteExecutionReceiptBuilderOpt{
+				utils.WithResults(1),
+				utils.WithChunks(1),
+				utils.WithCopies(1),
+			},
+			staked: false, // unstaked
+			msg:    "1 block, 1 result, 1 chunk, no duplicate, unstaked",
+		},
+		{
+			blockCount: 1,
+			ops: []utils.CompleteExecutionReceiptBuilderOpt{
+				utils.WithResults(5),
+				utils.WithChunks(5),
+				utils.WithCopies(1),
+			},
+			staked: true,
+			msg:    "1 block, 5 result, 5 chunks, no duplicate, staked",
+		},
+		{
+			blockCount: 10,
+			ops: []utils.CompleteExecutionReceiptBuilderOpt{
+				utils.WithResults(5),
+				utils.WithChunks(5),
+				utils.WithCopies(2),
+			},
+			staked: true,
+			msg:    "10 block, 5 result, 5 chunks, 1 duplicates, staked",
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf(tc.msg), func(t *testing.T) {
+			withConsumers(t, tc.staked, tc.blockCount, func(
+				blockConsumer *blockconsumer.BlockConsumer,
+				chunkConsumer *chunkconsumer.ChunkConsumer,
+				blocks []*flow.Block,
+				wg *sync.WaitGroup) {
+
+				unittest.RequireCloseBefore(t, chunkConsumer.Ready(), time.Second, "could not start chunk consumer")
+				unittest.RequireCloseBefore(t, blockConsumer.Ready(), time.Second, "could not start block consumer")
+
+				for i := 0; i < len(blocks); i++ {
+					// consumer is only required to be "notified" that a new finalized block available.
+					// It keeps track of the last finalized block it has read, and read the next height upon
+					// getting notified as follows:
+					blockConsumer.OnFinalizedBlock(&model.Block{})
```

Comment:
[The interface for HotStuff events](https://github.com/onflow/flow-go/blob/1e0a30e7a9118aad96a706639e37ec955f562542/consensus/hotstuff/consumer.go#L16) doesn't guarantee only-once delivery of events. If we receive a duplicate event here, the consumer will get out of sync with the actual finalized height.

Reply:
@jordanschalm thanks for the flag, but it does not pose a problem. The implementation of `OnFinalizationBlock` for the consumer just notifies it that _there may be a new block_, but it _does not_ change the internal state of its reader (e.g., finalized height). 

The call to `OnFinalizedBlock` here is routed to the following snippet, which tries to read the next job, and if it is not available, it gracefully returns the function: 
https://github.com/onflow/flow-go/blob/0c58b1831cfdba0ddee48dbc4e3c07b75b9cb9df/module/jobqueue/consumer.go#L265-L270

Now, who can write to the `jobs` here? `jobs` a wrapper around the `storage.Blocks` that is solely updated by the `follower` engine (in this case of a verification node):
https://github.com/onflow/flow-go/blob/2e696f1f96f6c0e45926b4fa06af8486420a35f9/engine/verification/assigner/blockconsumer/reader.go#L26-L44

So, regardless of how many times the repetition happens, the consumer stays synced with the block height and progresses _only if it is notified_ and _there is a block at the next height_.

To support this observation, I added an `eventRepetition` to the test cases that repeats notifying the consumer more than once for each finalized block, and it passes the tests: https://github.com/onflow/flow-go/pull/573/commits/2789256aa9fe54999774b55835578b764ce1bc70

Reply:
👌  Thanks for the explanation and tests